Hakawai/HKWMentionsPluginV2.m

```diff
@@ -0,0 +1,1590 @@
+//  the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+//  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//
+
+#import "HKWMentionsPluginV2.h"
+
+#import "HKWCustomAttributes.h"
+#import "HKWRoundedRectBackgroundAttributeValue.h"
+
+#import "HKWTextView.h"
+#import "HKWTextView+TextTransformation.h"
+#import "HKWTextView+Extras.h"
+#import "HKWTextView+Plugins.h"
+
+#import "HKWMentionsAttribute.h"
+
+#import "_HKWMentionsStartDetectionStateMachine.h"
+#import "_HKWMentionsCreationStateMachineV2.h"
+#import "_HKWMentionsCreationStateMachineV1.h"
+#import "_HKWMentionsCreationStateMachine.h"
+
+#import "_HKWMentionsPrivateConstants.h"
+
+@interface HKWMentionsPluginV2 () <HKWMentionsCreationStateMachineProtocol>
+
+// State properties
+@property (nonatomic) HKWMentionsState state;
+
+/// Whether or not initial setup has been performed. If the text view is the first responder already when the plug-in is
+/// set, initial setup should be performed immediately. However, if it's not, setup should be deferred until editing is
+/// going to begin, so that the text view doesn't snatch first-responder status from a different control.
+@property (nonatomic) BOOL initialSetupPerformed;
+
+/*!
+ Indicates that all calls to \c textViewDidChangeSelection: should be ignored. This is used when custom transformations
+ on the text view's text are desired, which would result in the text view's selection range temporarily changing, but
+ should not actually indicate the cursor being manually moved or text being selected.
+ */
+@property (nonatomic) BOOL suppressSelectionChangeNotifications;
+
+/*!
+ Indicates that the next call to \c textViewDidChangeSelection: should be ignored. This property is a fix for an issue
+ where the delegate method is called twice whenever a single character is deleted, each time with a different range,
+ screwing up the state machine.
+
+ \warning Do not use this to suppress notifications unless absolutely necessary. Instead, in many cases where text is
+ programmatically replaced, it is more correct to update the \c previousSelectionRange and \c previousTextLength
+ properties, preventing the spurious selection change detection machinery from getting into a bad state.
+ */
+@property (nonatomic) BOOL nextSelectionChangeShouldBeIgnored;
+
+/*!
+ Indicates that the next call to \c textView:shouldChangeTextInRange:replacementText: should be ignored. This property
+ is a fix for an issue where, when a suggestion-based keyboard suggestion is deleted (e.g. for the Chinese keyboard),
+ the delegate method is fired requesting the suggestion be deleted, and then immediately again with a spurious
+ insertion.
+ */
+@property (nonatomic) BOOL nextInsertionShouldBeIgnored;
```

Comment:
looks like they are not actually being used, can we remove?

Reply:
yes removed a lot more